### PR TITLE
[NPU] add index_sample_op_npu and tests

### PR DESCRIPTION
**Pull request by ronny1996**
### PR types

Others
### PR changes

OPs
### Describe

add index_sample_op_npu and tests
image

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/npu/test_index_sample_op_npu.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+from op_test import OpTest
+import paddle
+import paddle.fluid as fluid
+
+paddle.enable_static()
+
+
+@unittest.skipIf(not paddle.is_compiled_with_npu(),
+                 "core is not compiled with NPU")
```
Comment (qili93) [inline python/paddle/fluid/tests/unittests/npu/test_index_sample_op_npu.py]: 秋良在这个PR https://github.com/PaddlePaddle/Paddle/pull/34240 里面修改了，如果这个PR在秋良的PR之后合入的话，可以把这里的skipIf代码去掉，后面的代码也是。

Comment (ronny1996) [reply python/paddle/fluid/tests/unittests/npu/test_index_sample_op_npu.py]: 已更新

**Review by qili93 — APPROVED** [review]
LGTM

File: paddle/fluid/operators/index_sample_op_npu.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                            const Tensor* index, const Tensor* out_grad,
+                            Tensor* x_grad) {
+  auto index_dims = index->dims();
+  auto input_dims = x_grad->dims();
+  auto batch_size = input_dims[0];
+  auto index_length = index_dims[1];
+
+  std::vector<IndexT> scatter_index_vec;
+  std::vector<IndexT> index_vec;
+  framework::TensorToVector(*index, dev_ctx, &index_vec);
```
Comment (zhiqiu) [inline paddle/fluid/operators/index_sample_op_npu.cc:74]: Host-Device copy would make poor performance, is there any better way to implement this op?

**Review by zhiqiu — APPROVED** [review]
LGTM for shareDataWith

**Review by zhangting2020 — APPROVED** [review]

**Pull request merged**